### PR TITLE
`ultralytics 8.3.12` SAM and SAM2 multi-point prompts

### LINKED ISSUE #15716: Multiple points in SAM?
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hi Ultralytics!
It seems that the SAM now cannot accept points prompt that more than one, as the original meta sam can do so.
Personally, I struggled with the points prompt for a long time until I change to the libary of meta's.
Now I think I have time to help these enhencements and PR.
```python
from ultralytics import SAM

# Load a model
model = SAM("sam_b.pt")

# Segment with bounding box prompt
model("ultralytics/assets/zidane.jpg", bboxes=[439, 437, 524, 709])

# Segment with points prompt
model("ultralytics/assets/zidane.jpg", points=[900, 370], labels=[1])
```
change to
```python
from ultralytics import SAM

# Load a model
model = SAM("sam_b.pt")

# Segment with bounding box prompt
model("ultralytics/assets/zidane.jpg", bboxes=[439, 437, 524, 709])

# Segment with points prompt
model("ultralytics/assets/zidane.jpg", points=[[900, 370],[100,200],[450,400]], labels=[1,0,1])
```
Ref: [Ultralytics Doc](https://docs.ultralytics.com/models/sam/#how-can-i-use-the-segment-anything-model-sam-for-image-segmentation)
Expect behavior:
image

### Use case

People can use the SAM more easily and have more moethod to get the mask.

### Additional

Yes I'd like to help by submitting a PR!
But I am not sure if these enhencements were done or I am loss with the doc.
Feel free to tell me the details of it.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #16705: Negative points seems not to be used in any SAM model
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

Negative points seems to be considered as positive point in SAM model.

Here is an example with 1 positive point and 2 negative

points: [[     704.56      1315.8] [     848.99      1331.6]  [      912.4      1271.7]]
labels: [          1           0           0]

Capture d’écran 2024-10-05 à 11 31 18

I tried all sam models in this list, and the issue is present for all of them :  

### Environment

Ultralytics 8.3.4 🚀 Python-3.8.0 torch-2.2.2 CPU (Apple M3 Pro)
Setup complete ✅ (11 CPUs, 36.0 GB RAM, 176.1/926.4 GB disk)
None

### Minimal Reproducible Example

from ultralytics import SAM

model = SAM("sam2_t.pt")
results = model("./baloons.jpeg", points=[[150, 150], [100, 100]], labels=[1, 0])

### Additional

I saw a recent fix was released in 8.3.4:

https://github.com/ultralytics/ultralytics/pull/16642

but it does not seems to work

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/15716

The PR allows users to pass multiple points and labels as prompts for a single object.

To pass multiple points, users can use points and labels of shape `[num_objects, num_points, 2]` and `[num_objects, num_points]` respectively. The labels are still optional.

```python
In [2]: from ultralytics import SAM

In [3]: import torch

In [4]: model = SAM("mobile_sam.pt")

In [5]: points = torch.tensor([[[325,400], [700,400]]]).to("cuda:0")

In [6]: labels = torch.ones(points.shape[:-1], device="cuda:0", dtype=torch.int32)

In [7]: result = model("ultralytics/assets/bus.jpg", points=points, labels=labels)

image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 1024x1024 1 0, 357.1ms
Speed: 20.0ms preprocess, 357.1ms inference, 21.7ms postprocess per image at shape (1, 3, 1024, 1024)
```

image

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancement of the `prompt_inference` method in the prediction module to support multi-point inputs and improve validation.

### 📊 Key Changes
- Modified `points` and `labels` parameters to support multiple point sets.
- Added an assertion to verify that the number of points matches the number of labels.
- Updated processing logic to handle multi-dimensional inputs and ensure consistency.

### 🎯 Purpose & Impact
- 🛠️ **Flexibility**: Allows users to input multiple point sets for more detailed predictions.
- 🛡️ **Validation**: New checks prevent mismatches, reducing potential errors.
- 📈 **User Experience**: Enhances usability by accommodating more complex input scenarios, benefiting both developers and end-users.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced segmentation capabilities with improved prompt handling in the Ultralytics model.

### 📊 Key Changes
- Updated documentation for the MobileSAM and SAM models, demonstrating multiple usage scenarios.
- Added support for predicting segments using multiple points, points per object, and negative prompts.
- Introduced assertions to ensure point-label correspondence.
- Version bump from 8.3.11 to 8.3.12.

### 🎯 Purpose & Impact
- **Enhanced Functionality**: Supports more complex segmentation tasks with diverse prompting options, improving usability for a wider range of scenarios.
- **Usability Improvements**: Clearer examples make it easier for new users to understand and use the feature, boosting user experience.
- **Consistency Checks**: Ensures correct input formats, reducing errors and improving reliability.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a PR to the `ultralytics/ultralytics` 🚀 repository! This automated response is here to guide you through our process, and an Ultralytics engineer will review your submission soon. Please go through the checklist below to ensure everything is in order:

- ✅ **Define a Purpose**: Make sure your PR description clearly explains the purpose and the benefits of supporting multi-point prompts. If you haven't already, link to any related [issues](https://github.com/ultralytics/ultralytics/issues).
- ✅ **Synchronize with Source**: Check that your PR is up-to-date with the `main` branch. If needed, use the 'Update branch' button or run `git pull` and `git merge main` to resolve any conflicts.
- ✅ **Ensure CI Checks Pass**: Confirm all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are successful. If any checks fail, please look into resolving them.
- ✅ **Update Documentation**: Update any relevant [documentation](https://docs.ultralytics.com) sections to reflect the new feature in the SAM module.
- ✅ **Add Tests**: If applicable, add or modify tests to ensure coverage for the new multi-point support functionality.
- ✅ **Sign the CLA**: Ensure you've signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by commenting "I have read the CLA Document and I sign the CLA" if this is your first PR.
- ✅ **Minimize Changes**: Focus your changes to the minimum required for the enhancement. As Bruce Lee wisely put it, _"Hack away the unessential."_

For more detailed guidance, please visit our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Feel free to drop a comment if you have questions or need further assistance. Thank you for contributing to Ultralytics! 🚀

**Comment by Jordan-Pierce** [comment]
You're fast @Y-T-G!  This will also address #15716

**Comment by glenn-jocher** [comment]
@Y-T-G dont' think this makes sense:

- Why create a 3d Tensor and then slice it to 2d instead of creating a 2D tensor in the first place?
- Having to send to correct device will just create bug reports in the future
- Current points are a list and now we are creating a torch Tensor that is susceptible to being on the wrong device, I don' think users will understand this (I don't even understand it).
- No tests
- No docs updates for this use case

**Comment by Y-T-G** [comment]
**3D points and 2D labels**

```python
from ultralytics import SAM
model = SAM("mobile_sam.pt")
points = [[[325,400], [700,400]]]
labels = [[1,1]]
results  = model("ultralytics/assets/bus.jpg", points=points, labels=labels)
```
image

**2D points and 1D labels**

```python
from ultralytics import SAM
model = SAM("mobile_sam.pt")
points = [[325,400], [700,400]]
labels = [1,1]
results  = model("ultralytics/assets/bus.jpg", points=points, labels=labels)
```
image

**Comment by Y-T-G** [comment]
@glenn-jocher I made the required changes and added the tests. It should retain the functionality of the old approach along with the new approach.

**Review by RuellePaul — APPROVED** [review]

**Comment by glenn-jocher** [comment]
@Y-T-G looks good now! Thanks for the changes.

**Comment by DifferentialityDevelopment** [comment]
Thank you for implementing this! Super excited about this! I had to hack something together that doesn't work quite well just to be able to use SAM2 with multiple points, now I can use it natively.

**Comment by DifferentialityDevelopment** [comment]
multiple points seem to work with mobile_sam but not with SAM2, when passing the 3D list points [num_objects, num_points, 2] and 2D list labels [num_objects, num_points] to a SAM2 model I get the error
Tensors must have same number of dimensions: got 4 and 3

**Comment by Laughing-q** [comment]
@DifferentialityDevelopment Thanks for the reporting! I've submitted a PR https://github.com/ultralytics/ultralytics/pull/16894 to make it work with SAM2 models. :)
